### ISSUE TITLE
Cannot log in with current cookies after long session.

### ISSUE THREAD

**Issue by benshalev849** [PowerDNS-Admin/PowerDNS-Admin#1439]
I have setup the powerdns in runtime, after a while if I don't delete the cookies i get the following errors:

```
[2023-03-12 14:52:54,305] [_internal.py:224] INFO - 10.0.0.1 - - [12/Mar/2023 14:52:54] "GET / HTTP/1.1" 500 -
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2091, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.9/site-packages/werkzeug/middleware/proxy_fix.py", line 187, in __call__
    return self.app(environ, start_response)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2076, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/local/lib/python3.9/site-packages/flask/app.py", line 2072, in wsgi_app
    ctx.push()
  File "/usr/local/lib/python3.9/site-packages/flask/ctx.py", line 434, in push
    self.session = session_interface.open_session(self.app, self.request)
  File "/usr/local/lib/python3.9/site-packages/flask_session/sessions.py", line 535, in open_session
    if saved_session and saved_session.expiry <= datetime.utcnow():
TypeError: '<=' not supported between instances of 'NoneType' and 'datetime.datetime'
[2023-03-12 14:52:54,576] [_internal.py:224] INFO - 10.0.0.1 - - [12/Mar/2023 14:52:54] "GET /?__debugger__=yes&cmd=resource&f=style.css HTTP/1.1" 304 -
[2023-03-12 14:52:54,579] [_internal.py:224] INFO - 10.0.0.1 - - [12/Mar/2023 14:52:54] "GET /?__debugger__=yes&cmd=resource&f=debugger.js HTTP/1.1" 304 -
[2023-03-12 14:52:55,591] [_internal.py:224] INFO - 10.0.0.1 - - [12/Mar/2023 14:52:55] "GET /?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 304 -
[2023-03-12 14:52:55,664] [_internal.py:224] INFO - 10.0.0.1 - - [12/Mar/2023 14:52:55] "GET /?__debugger__=yes&cmd=resource&f=console.png HTTP/1.1" 304 -
```

This makes me unable to log in or access the DNS, the workaround i have found is either clear all cookies or accessing in incognito.

**Comment by AzorianMatt** [comment]
@benshalev849 did you by chance reset your database during some of your work? Now that the sessions are being stored in the database by default, I could see where this might have happened.

**Comment by subbink** [comment]
Noticed the same today. I solved it by clearing the sessions table in the database, since it was only a test setup.

**Comment by AzorianMatt** [comment]
@subbink this has me wondering if this is a newly introduced issue from the recent PR that integrated SQLAlchemy for session storage.

**Comment by benshalev849** [comment]
@AzorianMatt I think this is possible due to me having a SESSION_TYPE as sqlalchemy with a mariadb as a backend.

In addition it's an old DB with previous created tables (Since it's a prep env it doesn't have anything on it, but it was used a bit ago).

**Comment by MrGeneration** [comment]
Just wanted to say that I encountered the same issue in the following two scenarious:
- the session timed out without proper logging out (above error appears)
- logged in with valid session, updated to 0.4 and refreshed the page (above error appears)

Every time this issue appears removing the site's cookies is enough.
The issue does not seem to occur when you're logging out cleanly.

This was a migrated installation via docker with MariaDB backend.

**Comment by AzorianMatt** [comment]
@MrGeneration I haven't replicated the problem yet myself but it's starting to sound like it's an issue directly tied to the changeover for session storage, which would make complete sense as it changed with this new release.

Ergo, after upgrading, all existing sessions would become invalid and likely lead to this kind of exception even though it should be more properly handled in the library.

**Comment by tbe** [comment]
@AzorianMatt on a fresh install, the same issue happens. There was no upgrade from an old database schema.

**Comment by raunz** [comment]
Same here. I had long session lifetime settings, so didn't notice it earlier. By setting short lifetime same errors occur and instance if giving http/500.
I'll try to debug it too... Looks like a non backend generic flask session code being used.
```
[2023-03-13 11:46:45,871] [app.py:1449] ERROR - Exception on /login [GET]
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2072, in wsgi_app
    ctx.push()
  File "/usr/lib/python3.10/site-packages/flask/ctx.py", line 434, in push
    self.session = session_interface.open_session(self.app, self.request)
  File "/usr/lib/python3.10/site-packages/flask_session/sessions.py", line 535, in open_session
    if saved_session and saved_session.expiry <= datetime.utcnow():
TypeError: '<=' not supported between instances of 'NoneType' and 'datetime.datetime'
[2023-03-13 11:46:45 +0000] [9] [ERROR] Error handling request /login
Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2072, in wsgi_app
    ctx.push()
  File "/usr/lib/python3.10/site-packages/flask/ctx.py", line 434, in push
    self.session = session_interface.open_session(self.app, self.request)
  File "/usr/lib/python3.10/site-packages/flask_session/sessions.py", line 535, in open_session
    if saved_session and saved_session.expiry <= datetime.utcnow():
TypeError: '<=' not supported between instances of 'NoneType' and 'datetime.datetime'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/gunicorn/workers/sync.py", line 136, in handle
    self.handle_request(listener, req, client, addr)
  File "/usr/lib/python3.10/site-packages/gunicorn/workers/sync.py", line 179, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2091, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/lib/python3.10/site-packages/werkzeug/middleware/proxy_fix.py", line 187, in __call__
    return self.app(environ, start_response)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 2076, in wsgi_app
    response = self.handle_exception(e)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 1432, in handle_exception
    server_error = self.ensure_sync(handler)(server_error)
  File "/app/powerdnsadmin/routes/base.py", line 34, in handle_internal_server_error
    return render_template('errors/500.html', code=500, message=e), 500
  File "/usr/lib/python3.10/site-packages/flask/templating.py", line 153, in render_template
    ctx.app.update_template_context(context)
  File "/usr/lib/python3.10/site-packages/flask/app.py", line 750, in update_template_context
    context.update(func())
  File "/usr/lib/python3.10/site-packages/flask_login/utils.py", line 407, in _user_context_processor
    return dict(current_user=_get_user())
  File "/usr/lib/python3.10/site-packages/flask_login/utils.py", line 372, in _get_user
    current_app.login_manager._load_user()
  File "/usr/lib/python3.10/site-packages/flask_login/login_manager.py", line 362, in _load_user
    user_id = session.get("_user_id")
AttributeError: 'NoneType' object has no attribute 'get'
```

**Comment by hellEVIL-13** [comment]
me to

Mar 13 16:13:09 webdns gunicorn[14646]: Traceback (most recent call last):
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/app.py", line 2072, in wsgi_app
Mar 13 16:13:09 webdns gunicorn[14646]:     ctx.push()
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/ctx.py", line 434, in push
Mar 13 16:13:09 webdns gunicorn[14646]:     self.session = session_interface.open_session(self.app, self.request)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask_session/sessions.py", line 535, in open_session
Mar 13 16:13:09 webdns gunicorn[14646]:     if saved_session and saved_session.expiry <= datetime.utcnow():
Mar 13 16:13:09 webdns gunicorn[14646]: TypeError: '<=' not supported between instances of 'NoneType' and 'datetime.datetime'
Mar 13 16:13:09 webdns gunicorn[14646]: [2023-03-13 16:13:09 +0300] [14646] [ERROR] Error handling request /
Mar 13 16:13:09 webdns gunicorn[14646]: Traceback (most recent call last):
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/app.py", line 2072, in wsgi_app
Mar 13 16:13:09 webdns gunicorn[14646]:     ctx.push()
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/ctx.py", line 434, in push
Mar 13 16:13:09 webdns gunicorn[14646]:     self.session = session_interface.open_session(self.app, self.request)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask_session/sessions.py", line 535, in open_session
Mar 13 16:13:09 webdns gunicorn[14646]:     if saved_session and saved_session.expiry <= datetime.utcnow():
Mar 13 16:13:09 webdns gunicorn[14646]: TypeError: '<=' not supported between instances of 'NoneType' and 'datetime.datetime'
Mar 13 16:13:09 webdns gunicorn[14646]: During handling of the above exception, another exception occurred:
Mar 13 16:13:09 webdns gunicorn[14646]: Traceback (most recent call last):
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/gunicorn/workers/sync.py", line 136, in handle
Mar 13 16:13:09 webdns gunicorn[14646]:     self.handle_request(listener, req, client, addr)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/gunicorn/workers/sync.py", line 179, in handle_request
Mar 13 16:13:09 webdns gunicorn[14646]:     respiter = self.wsgi(environ, resp.start_response)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/app.py", line 2091, in __call__
Mar 13 16:13:09 webdns gunicorn[14646]:     return self.wsgi_app(environ, start_response)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/werkzeug/middleware/proxy_fix.py", line 187, in __call__
Mar 13 16:13:09 webdns gunicorn[14646]:     return self.app(environ, start_response)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/app.py", line 2076, in wsgi_app
Mar 13 16:13:09 webdns gunicorn[14646]:     response = self.handle_exception(e)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/app.py", line 1432, in handle_exception
Mar 13 16:13:09 webdns gunicorn[14646]:     server_error = self.ensure_sync(handler)(server_error)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/powerdnsadmin/routes/base.py", line 34, in handle_internal_server_error
Mar 13 16:13:09 webdns gunicorn[14646]:     return render_template('errors/500.html', code=500, message=e), 500
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/templating.py", line 153, in render_template
Mar 13 16:13:09 webdns gunicorn[14646]:     ctx.app.update_template_context(context)
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask/app.py", line 750, in update_template_context
Mar 13 16:13:09 webdns gunicorn[14646]:     context.update(func())
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask_login/utils.py", line 407, in _user_context_processor
Mar 13 16:13:09 webdns gunicorn[14646]:     return dict(current_user=_get_user())
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask_login/utils.py", line 372, in _get_user
Mar 13 16:13:09 webdns gunicorn[14646]:     current_app.login_manager._load_user()
Mar 13 16:13:09 webdns gunicorn[14646]:   File "/opt/web/powerdns-admin/venv/lib/python3.9/site-packages/flask_login/login_manager.py", line 362, in _load_user
Mar 13 16:13:09 webdns gunicorn[14646]:     user_id = session.get("_user_id")
Mar 13 16:13:09 webdns gunicorn[14646]: AttributeError: 'NoneType' object has no attribute 'get'
Mar 13 16:13:10 webdns gunicorn[14646]: [2023-03-13 16:13:10,011] [app.py:1449] ERROR - Exception on / [GET]

**Comment by tbe** [comment]
This is a bug in Flask-Session: https://github.com/fengsp/flask-session/issues/151

But there is hope, https://github.com/christopherpickering/flask-session2 has not the same issue, as it checks the session object before using it, and it should be a simple drop in replacement @AzorianMatt

**Comment by AzorianMatt** [comment]
@tbe thank you for researching this. I will discuss this further with the contribution team and look at getting It upgraded with the drop in replacement.

**Comment by raunz** [comment]
Currently used [flask-session](https://github.com/fengsp/flask-session) has been prone to this error since 2017. According to the project issues, there are several complaints about the same issue and overall it's related how the application destroys session. Also some concurrency has a part in this (multiple pod race condition).
The currently maintained drop-in replacement [flask-session2](https://github.com/christopherpickering/flask-session2) has merged the fix to handle session data with NULL expire value, but not the root cause of avoiding inserting NULL values in first place.

Somehow a query with insert NULL will be issued. I suspect a race condition with using `session.clear()` in PDA code.
```
INSERT INTO sessions (session_id, data, expiry) 
VALUES ('session:f933fc61-0bfc-4d13-b2ee-ca2b4ce6388e', 'U\0\0\0\0\0\0\0}_csrf_token@e9242866c9a6e8eaeacbc4258deef1f657c7e7db5d6f272bdb47d27c..s.', NULL)
```

I made a simple test by removing excess line from clear_session() function, by letting flask-session native logout_user() do all the session invalidation tricks.  Fun fact - the sessions don't get broken anymore. I can't reproduce the situation any more, as all session entries in backend database have proper "not NULL" values.

```
diff --git a/powerdnsadmin/routes/index.py b/powerdnsadmin/routes/index.py
index c145f6c..28b4e03 100644
--- a/powerdnsadmin/routes/index.py
+++ b/powerdnsadmin/routes/index.py
@@ -526,7 +526,6 @@ def clear_session():
     session.pop('google_token', None)
     session.pop('authentication_type', None)
     session.pop('remote_user', None)
-    session.clear()
     logout_user()
```
I can't say, what was the initial purpose of adding session.clear() to the PDA code. Maybe this breaks something in the other end, but using both, session.clear() and flask sessions logout_user(), isn't a great combination. As in previous releases only the 'filesystem' session store was being used, the concurrency never triggered an issue. According to the issues of flask-session, it's common to all centralized backends (sql, memcache, redis...).

**Comment by tbe** [comment]
To @raunz comment: I have experienced this issue with expired session. So, i would assume that here may be more then one issue with the same result

**Comment by raunz** [comment]
@tbe 
> I have experienced this issue with expired session. So, i would assume that here may be more then one issue with the same result

Did You clear the backend from stale "NULL" expiry sessions?  The change to remove session.clear() doesn't patch the TypeError occuring. I cleared the database with the following and restarted PDA with patched powerdnsadmin/routes/index.py code.
```
 delete from sessions where expiry is null; 
 or 
 truncate sessions;
```

**Comment by tbe** [comment]
@raunz I started with an empty database (new installation) and my first session expired after 10 minutes, resulting in 500 with the same error. 

Also, the session is looked up by ID, so even if there are entries with NULL values in the db, it should not impact the result

**Comment by AzorianMatt** [comment]
@raunz Do you have some capacity to do a PR for this change to flask-session2 by chance?

**Comment by nkukard** [comment]
> @raunz Do you have some capacity to do a PR for this change to flask-session2 by chance?

On a side note .... (don't take me seriously though, I'm by far a Flask expert) ... I looked into this issue a few days ago to see if I could get Flask-Session2 switched out and it required a newer version of Flask. Flask is pinned to 2.1.3 in requirements.txt, Flask-Session2 requires 2.2.2.

I then tried to update Flask to 2.2.2 but that appeared to break quite a number more things. We may need to update to Flask 2.2.2 to use Flask-Session2.

Here is the dependency issue output log on Ubuntu 22.04... (with just the Flask-Session2 change)
```
Collecting Flask-Session2
  Downloading Flask_Session2-1.3.0-py3-none-any.whl (14 kB)
  Downloading Flask_Session2-1.2.0-py3-none-any.whl (11 kB)
  Downloading Flask_Session2-1.1.0-py3-none-any.whl (10 kB)
  Downloading Flask_Session2-1.0.0-py3-none-any.whl (9.8 kB)
INFO: pip is looking at multiple versions of psycopg2 to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of zxcvbn to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of rcssmin to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of zipp to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of werkzeug to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of webcolors to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of rjsmin to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of requests to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of qrcode to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pytz to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pytimeparse to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of python3-saml to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of python-ldap to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pytest to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyotp to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyasn1 to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of passlib to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of mysqlclient to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of lima to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of itsdangerous to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of gunicorn to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask-session-captcha to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of cssmin to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of cryptography to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of configobj to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of cffi to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of certifi to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of bravado-core to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of bcrypt to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of sqlalchemy to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of pyyaml to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of jinja2 to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask-seasurf to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask-sslify to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask-sqlalchemy to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask-migrate to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask-mail to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of <Python from Requires-Python> to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask-login to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of flask-assets to determine which version is compatible with other requirements. This could take a while.
INFO: pip is looking at multiple versions of authlib to determine which version is compatible with other requirements. This could take a while.
ERROR: Cannot install -r requirements.txt (line 2), -r requirements.txt (line 23), -r requirements.txt (line 3), -r requirements.txt (line 4), -r requirements.txt (line 5), -r requirements.txt (line 6), -r requirements.txt (line 7), -r requirements.txt (line 8), -r requirements.txt (line 9) and Flask==2.1.3 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested Flask==2.1.3
    flask-assets 2.0 depends on Flask>=0.8
    flask-login 0.6.2 depends on Flask>=1.0.4
    flask-mail 0.9.1 depends on Flask
    flask-migrate 2.5.3 depends on Flask>=0.9
    flask-sqlalchemy 2.5.1 depends on Flask>=0.10
    flask-sslify 0.1.5 depends on Flask
    flask-seasurf 1.1.1 depends on Flask
    flask-session-captcha 1.3.0 depends on Flask
    flask-session2 1.3.1 depends on Flask<3.0.0 and >=2.2.2
    The user requested Flask==2.1.3
    flask-assets 2.0 depends on Flask>=0.8
    flask-login 0.6.2 depends on Flask>=1.0.4
    flask-mail 0.9.1 depends on Flask
    flask-migrate 2.5.3 depends on Flask>=0.9
    flask-sqlalchemy 2.5.1 depends on Flask>=0.10
    flask-sslify 0.1.5 depends on Flask
    flask-seasurf 1.1.1 depends on Flask
    flask-session-captcha 1.3.0 depends on Flask
    flask-session2 1.3.0 depends on Flask<3.0.0 and >=2.2.2
    The user requested Flask==2.1.3
    flask-assets 2.0 depends on Flask>=0.8
    flask-login 0.6.2 depends on Flask>=1.0.4
    flask-mail 0.9.1 depends on Flask
    flask-migrate 2.5.3 depends on Flask>=0.9
    flask-sqlalchemy 2.5.1 depends on Flask>=0.10
    flask-sslify 0.1.5 depends on Flask
    flask-seasurf 1.1.1 depends on Flask
    flask-session-captcha 1.3.0 depends on Flask
    flask-session2 1.2.0 depends on Flask<3.0.0 and >=2.2.2
    The user requested Flask==2.1.3
    flask-assets 2.0 depends on Flask>=0.8
    flask-login 0.6.2 depends on Flask>=1.0.4
    flask-mail 0.9.1 depends on Flask
    flask-migrate 2.5.3 depends on Flask>=0.9
    flask-sqlalchemy 2.5.1 depends on Flask>=0.10
    flask-sslify 0.1.5 depends on Flask
    flask-seasurf 1.1.1 depends on Flask
    flask-session-captcha 1.3.0 depends on Flask
    flask-session2 1.1.0 depends on Flask<3.0.0 and >=2.2.2
    The user requested Flask==2.1.3
    flask-assets 2.0 depends on Flask>=0.8
    flask-login 0.6.2 depends on Flask>=1.0.4
    flask-mail 0.9.1 depends on Flask
    flask-migrate 2.5.3 depends on Flask>=0.9
    flask-sqlalchemy 2.5.1 depends on Flask>=0.10
    flask-sslify 0.1.5 depends on Flask
    flask-seasurf 1.1.1 depends on Flask
    flask-session-captcha 1.3.0 depends on Flask
    flask-session2 1.0.0 depends on Flask<3.0.0 and >=2.2.2

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/topics/dependency-resolution/#dealing-with-dependency-conflicts
```

**Comment by AzorianMatt** [comment]
@nkukard go figure, the dependency lock gets us again! I suppose we'll need to have another discussion about our short-term vs long-term approach here because I imagine such a big upgrade will destabilize the software more while I'm trying to increase stability. Thank you for checking this out!

**Comment by subbink** [comment]
Would it be an idea to stop using the database for session storage, if that solves the issue?

**Comment by AzorianMatt** [comment]
@subbink possibly, but I would hate to go and flip the default yet again, as I know the original change was brought about just to support multi-homed deployments.

I would like to know how to reproduce this issue, because I have yet to run into it myself. Granted, I have been running my local test environment on SQLite, although I wouldn't think that matters.

**Comment by nkukard** [comment]
> @subbink possibly, but I would hate to go and flip the default yet again, as I know the original change was brought about just to support multi-homed deployments.
> 
> I would like to know how to reproduce this issue, because I have yet to run into it myself. Granted, I have been running my local test environment on SQLite, although I wouldn't think that matters.

I can reproduce (at least the session timeout exception case) using MySQL and PostgreSQL by logging into the UI, then just leaving the window open for maybe an hour or two (generally I get pulled away for some other thing I need to attend), when I get back to the UI I'm at a 500 error screen.

Happened 100% of the time. I think the above mention of removing session clear may fix  this particular issue though.

I can maybe test out if removing the session clear fixes this case, at least if it does its one step in the right  direction.

**Comment by raunz** [comment]
@AzorianMatt let's keep the default backend as sqlalchemy, but the "session.clear()" has to go from PDA code. As I noted above - the fault isn't the flask-session module, but how PDA uses flask-session functions.
To reproduce You need real-SQL backend (postgres or mysql) and following sequence:
1. Turn session lifetime to for example 5 minutes. 
2. Log out/Log in again and leave PDA idle in browser (session expire timer modal will pop up in 4 minutes). Still do nothing
3. browser will be redirected to login page, but you can enjoy the view of error 500 and inoperable application (session table contains now rows with expire=NULL).

I even tried to move session.clear() line below logout_user(), but no help. Still getting 500.
```
-    session.clear()
     logout_user()
+    session.clear()
```
To get rid of the problem: remove [session.clear() line](https://github.com/PowerDNS-Admin/PowerDNS-Admin/blob/master/powerdnsadmin/routes/index.py#L529) and clean up the database with `delete from sessions where expiry is null;
`

**Comment by AzorianMatt** [comment]
@raunz FYI, this can also be reproduced on SQLite back-end as it just happened to me for the first time on my local venv setup.

Would you mind submitting a PR for this and I'll merge it quickly?

**Comment by subbink** [comment]
it seems to me that this issue has been fixed with the MR #1461

**Comment by AzorianMatt** [comment]
I'm happy to report that some of the discussed and long known issues involving the flask-session dependency, seem to have been solved in the latest update of `0.6.0` which I am testing in the `dev` branch now.

https://github.com/PowerDNS-Admin/PowerDNS-Admin/pull/1748
